File: divisor-java-springboot-webflux/src/main/java/my/test/divisor/java/springboot/webflux/domain/service/DivisorService.java
package my.test.divisor.java.springboot.webflux.domain.service;

import my.test.divisor.java.springboot.webflux.domain.model.FilteringOption;
import reactor.core.publisher.Flux;

import java.util.Optional;
import java.util.function.Predicate;

/**
 * Service to find divisors for provided
 */
public class DivisorService {

    /**
     * predicate to check that the number is even
     */
    private static Predicate<Integer> isEven = number -> number % 2 == 0;

    /**
     * predicate to check that the number is odd
     */
    private static Predicate<Integer> isOdd = isEven.negate();

    /**
     * lower end for the divisor range
     */
    private final int divisorRangeLowerEnd;

    /**
     * upper end for the divisor range
     */
    private final int divisorRangeUpperEnd;

    /**
     * Creates a new instance of DivisorService with provided parameters for divisor range.
     *
     * @param divisorRangeLowerEnd lower end for the divisor range
     * @param divisorRangeUpperEnd upper end for the divisor range
     */
    public DivisorService(int divisorRangeLowerEnd, int divisorRangeUpperEnd) {
        this.divisorRangeLowerEnd = divisorRangeLowerEnd;
        this.divisorRangeUpperEnd = divisorRangeUpperEnd;
    }

    /**
     * Finds stream of divisors for the provided dividend, filters the resulting stream of divisors if the numberValue
     * is not <dd>null</dd>
     *
     * @param numberValue     dividend
     * @param filteringOption defines what filtering should be applied on the resulting stream of divisors
     * @return stream of found divisors, empty stream if &quot;numberValue&quot; is <dd>null</dd>
     */
    public Flux<Integer> findDivisor(Integer numberValue, Optional<FilteringOption> filteringOption) {
        if (numberValue == null) {
            return Flux.empty();
        }
        int count = divisorRangeUpperEnd - divisorRangeLowerEnd + 1;
        Flux<Integer> divisorStream = Flux.range(divisorRangeLowerEnd, count)
                .filter(candidate -> numberValue % candidate == 0);
        return filterDivisors(divisorStream, filteringOption);
    }

    /**
     * Applies filtering of the provided stream. A proper predicate {@link #isEven} or {@link #isOdd} is applied based
     * on the &quot;filteringOption&quot; parameter value. The source stream is returned if &quot;filteringOption&quot;
     * is empty
     *
     * @param divisorStream   source stream to apply filtering
     * @param filteringOption difines what filtering should be applied
     * @return source stream if the &quot;filteringOption&quot; parameter is empty or the filtered stream otherwise
     */
    private Flux<Integer> filterDivisors(Flux<Integer> divisorStream, Optional<FilteringOption> filteringOption) {
        if (filteringOption.isEmpty()) {
            return divisorStream;
        }
        Predicate<Integer> predicateToApply = filteringOption.get() == FilteringOption.ODD ? isOdd : isEven;
        return divisorStream.filter(predicateToApply);
    }
}

File: divisor-kotlin-springboot-mvc/src/main/kotlin/my/test/divisor/kotlin/springboot/mvc/domain/service/DivisorService.kt
package my.test.divisor.kotlin.springboot.mvc.domain.service

import my.test.divisor.kotlin.springboot.mvc.domain.model.FilteringOption

/**
 * Service to find divisors for provided
 */
class DivisorService(
    private val divisorRangeLowerEnd: Int,
    private val divisorRangeUpperEnd: Int
) {

    /**
     * Finds list of divisors for the provided dividend number, filtered if the numberValue is not [null].
     *
     * @param numberValue the dividend number to find the divisors for
     * @param filteringOption defines what filtering should be applied on the resulting list of divisors
     * @return list of found and optionally filtered divisors or empty list if [numberValue] is [null]
     */
    fun findDivisor(numberValue: Int, filteringOption: FilteringOption?): List<Int> =
        (divisorRangeLowerEnd..divisorRangeUpperEnd)
            .filter { numberValue % it == 0 }
            .let { divisors ->
                when (filteringOption) {
                    null -> divisors
                    FilteringOption.EVEN -> divisors.filter { it % 2 == 0 }
                    FilteringOption.ODD -> divisors.filter { it % 2 != 0 }
                }
            }
}